encrypt/aeadCipher.go

```diff
@@ -0,0 +1,376 @@
+	NonceSize        = 12
+	readBufferSize   = 0xFFFFF             // get 4k for buffer size
+	hkdfInfoTag      = []byte("ss-subkey") // at present, this tag is defined as ss-subkey
+	nullNonce        = make([]byte, NonceSize, NonceSize)
+	aeadCipherMethod = map[string]*aeadGenerator{
+		"chacha20-ietf-poly1305": {32, 32, newChaCha20IETFEncoder},
+		"aes-256-gcm":            {32, 32, newAESGCMEncoder},
+		"aes-192-gcm":            {24, 24, newAESGCMEncoder},
+		"aes-128-gcm":            {16, 16, newAESGCMEncoder},
+	}
+)
+
+type aeadGenerator struct {
+	keyLen  int
+	saltLen int
+	newAEAD func(key, salt []byte, keylen int) (cipher.AEAD, error)
+}
+
+// increase uesd for nonce increacement, will warpped when reach the maximum of length byte can reach
+func increase(src []byte) {
+	for i := range src {
+		src[i]++
+		if src[i] != 0 {
+			return
+		}
+	}
+}
+
+// generate the salt randomly fror user
+func genSalt(salt []byte) {
+	// if salt is NULL, then gen the salt randomly
+	// by default, salt is random generated for each invocation by writer
+	// reader should input the specific salt to hkdf key
+	switch len(salt) {
+	case 16, 24, 32:
+	default:
+		panic("error salt len")
+	}
+	//FIXME
+	n, err := io.ReadFull(rand.Reader, salt)
+	if err != nil || n != len(salt) {
+		panic("error salt gen")
+	}
+}
+
+// key-derivation function for key pre-handle
+func kdf(password string, keyLen int) []byte {
+	var b, prev []byte
+	h := md5.New()
+	for len(b) < keyLen {
+		h.Write(prev)
+		h.Write([]byte(password))
+		b = h.Sum(b)
+		prev = b[len(b)-h.Size():]
+		h.Reset()
+	}
+	return b[:keyLen]
+}
+
+type aeadCipher struct {
+	psk        []byte
+	salt       []byte
+	readNonce  []byte
+	writeNonce []byte
+	readBuffer []byte
+	saltSize   int
+	keySize    int
+	dataLen    int
+
+	enc     cipher.AEAD
+	dec     cipher.AEAD
+	genator func(key, salt []byte, keylen int) (cipher.AEAD, error)
+}
+
+// NewAEADCipher creates a cipher that can be used in Dial() etc.
+// Use cipher.Copy() to create a new cipher with the same method and password
+// to avoid the cost of repeated cipher initialization.
+func NewAEADCipher(method, password string) (c Cipher, err error) {
+	if password == "" {
+		return nil, ErrEmptyPassword
+	}
+
+	var genator *aeadGenerator
+	genator, ok := aeadCipherMethod[method]
+	if !ok {
+		return nil, ErrUnsupportedMethod
+	}
+
+	cipher := aeadCipher{
+		readNonce:  make([]byte, NonceSize, NonceSize),
+		writeNonce: make([]byte, NonceSize, NonceSize),
+		salt:       make([]byte, genator.saltLen, genator.saltLen),
+		psk:        kdf(password, genator.keyLen), // per handle the passwd
+		keySize:    genator.keyLen,
+		saltSize:   genator.saltLen,
+		genator:    genator.newAEAD,
+		readBuffer: make([]byte, readBufferSize),
+	}
+
+	return &cipher, nil
+}
+
+func (c *aeadCipher) KeySize() int          { return c.keySize }
+func (c *aeadCipher) InitBolckSize() int    { return c.saltSize }
+func (c *aeadCipher) EncryptorInited() bool { return c.enc == nil }
```

Comment:
The function name contradict its content==
just a remind==

Reply:
WIP